Add support for out-of-tree platform plugins

empyrical opened the pull request:
This pull request adds the ability for a platform developer to provide a `"haste"` key under the `"rnpm"` key in their `package.json` which allows the packager to pick up that platform's javascript files. The intent is to remove the need to have custom platforms hardcoded in. This is inspired by the `"jest": { "haste": {} }` key used by jest.

For example, React Native Dom would have an entry like:

```json
{
  "rnpm": {
    "haste": {
      "providesModuleNodeModules": [
        "react-native-dom"
      ],
      "platforms": [
        "dom"
      ]
    }
  }
}
```

Support for more keys (path blacklists perhaps?) could be added in the future.

This succeeds #20662, as per a discussion I had with @matthargett.

I've got an open discussion over here as well: https://github.com/react-native-community/discussions-and-proposals/issues/21

Test Plan:
----------
I've modified a local copy of RNDom to have the above example in its `package.json` in an unmodified default React Native app project, and a JS bundle was build successfully.

Release Notes:
--------------

[CLI] [ENHANCEMENT] [local-cli/core/findPlugins.js] - Modified the `findPlugins` function to handle this new `haste` field
[CLI] [ENHANCEMENT] [local-cli/core/index.js] - Removed the hardcoded platforms and switched to use output from `findPlugins`

empyrical commented:
@rozele: I had to change around the lines you modified in #20705 because it seems like `config.resolver.platforms` was always present with metro defaults, meaning `defaultConfig.getPlatforms()` was never reached even without a `metro.config.js`. I still have the `undefined` check there because I am not sure if there are still cases where it will be empty. I am not sure if this will break your use case or not.

empyrical commented on local-cli/core/findPlugins.js:
```diff
@@ -71,11 +98,12 @@ const findPluginInFolder = folder => {
         if (pkgJson) {
           commands = commands.concat(findPluginsInReactNativePackage(pkgJson));
```
I am not entirely sure what this means

matthargett commented:
@rozele @vincentriemer @captainnic I’m closing #20662 in favor of this approach.

react-linter commented:
<!--
  0 failure: 
  1 warning:  :exclamation: Big...
  
  
  DangerID: danger-id-default;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  :exclamation: Big PR - <i>This PR is extremely unlikely to get reviewed because it touches 868 lines.</i>
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

empyrical commented:
I've added a dummy platform in `RNTester/js/OutOfTreeTestPlatform` with the aim of testing RNCLI's support for bundling out of tree platforms.

Currently, there is just a test in `ContainerShip/scripts/run-ci-e2e-tests.sh` - I wanted to also have a platform suffix test in `jest/__tests__/hasteImpl-test.js`, but I need to be able to have `OutOfTreeTestPlatform`/`react-native-dummy` as a `devDependency` somehow for that to work. I do not predict this dummy module needing to be changed much - perhaps it could just be installed from npm, and added as another package that gets published in the publish script?

I currently have this module on NPM for testing purposes that I could transfer ownership over if FB devs are interested:

https://www.npmjs.com/package/react-native-dummy

matthargett commented on RNTester/js/OutOfTreeTestPlatform/Libraries/Components/ProgressViewIOS/ProgressViewIOS.dummy.js:
```diff
@@ -0,0 +1,46 @@
+
+class DummyProgressViewIOS extends React.Component {
+  render() {
+    return (
```
why not just have the render throw("Not implemented") or something similar? or just export `require('UnimplementedView');` like the other files?

matthargett commented on RNTester/js/OutOfTreeTestPlatform/Libraries/Components/SegmentedControlIOS/SegmentedControlIOS.dummy.js:
```diff
@@ -0,0 +1,46 @@
+class DummySegmentedControlIOS extends React.Component {
+  render() {
+    return (
+      <View style={[styles.dummy, this.props.style]}>
```
same question here -- why not just throw?

matthargett commented on RNTester/js/OutOfTreeTestPlatform/Libraries/Components/StatusBar/StatusBarIOS.dummy.js:
```diff
@@ -0,0 +1,13 @@
+
+'use strict';
+
+module.exports = {};
```
why empty here and not `require('UnimplementedView');` as elsewhere?

empyrical replied:
StatusBarIOS isn't a React component

https://github.com/facebook/react-native/blob/master/Libraries/Components/StatusBar/StatusBarIOS.ios.js

None if this code will be ran at all at any point - it is purely for the bundler's sake, so the actual code isn't *that* important. A lot of the shims, like `DummySegmentedControlIOS` are just copied over from Android shims in the RN codebase. Sorry for not making this clearer! I'll edit the main post

matthargett commented on RNTester/js/OutOfTreeTestPlatform/Libraries/Settings/Settings.dummy.js:
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+// Completely empty shim for the sake of bundling.
+const Settings = {};
```
if `require('UnimplementedView');` can't work here, let's inline the assignment as elsewhere.

empyrical commented:
I've run unit tests on my fork on a branch that's been rebased against #20854, and it turns out that `jest-haste-map` doesn't like two instances of this module being in its search paths at the same time, and it still seems to see it despite me adding it to all search path blacklists.

```
jest-haste-map: @providesModule naming collision:
  Duplicate module name: react-native-dummy
  Paths: /home/circleci/react-native/node_modules/react-native-dummy/package.json collides with /home/circleci/react-native/RNTester/js/OutOfTreeTestPlatform/package.json

This error is caused by a @providesModule declaration with the same name across two different files.
```

I honestly can't think of a way to get around this, so I will be trying this again as a separate pull request. If anyone wants to peek around at the code, it still exists at this branch:

https://github.com/empyrical/react-native/tree/haste-plugins-plus-test

jhampton commented:
@empyrical I _think_ have run across this before.  I was trying to do module substitution to emulate tree shaking for an out-of-tree platform.  My approach used the `getTransformModulePath` option in the rn-cli.config.js (I haven't found a ton of documentation, so the signature may have changed).  Basically this option is a _path_ to a _file_ that returns a _function_ that lets you hook into the module resolution and decide to replace the implementation wholesale.  I can provide a sample and basic walkthrough of what I was doing if you can DM me.

I'm by no means a Metro expert, but I've done a bit of hacking lately with it.

Cheers, and I'll take a look at your fork as well to see what I see.

-Jeff

kelset commented:
@rozele @vincentriemer I'd love to hear your feedback on this, seems solid - and thanks for adding the tests too!

vincentriemer commented:
I’ll try and take a look tonight 👌

vincentriemer commented:
Short of actually pulling the branch and testing it against rndom itself, LGTM 👍
